tools/clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -8013,6 +8013,54 @@ def err_hlsl_reorder_unsupported_stage : Error<
    "dx::MaybeReorderThread is unavailable in shader stage '%0' (requires 'raygeneration')">;
 def err_hlsl_hitobject_unsupported_stage : Error<
    "dx::HitObject is unavailable in shader stage '%0' (requires 'raygeneration', 'closesthit' or 'miss')">;
+
+// Linear Algebra Operations
+def err_hlsl_linalg_param_must_be_const : Error<
+  "'%0' must be a constant parameter">;
+def err_hlsl_linalg_vector_incorrect_type : Error<
+  "%0 is incorrect type, must be 16-bit or 32-bit 'unsigned int', "
+  "'signed int' or 'float'">;
+def err_hlsl_linalg_isunsigned_incorrect_for_given_type : Error<
+  "%0 must be %select{false|true}1 for a %2 vector type">;
+def err_hlsl_linalg_interpretation_value_incorrect : Error<
+  "%0 is an invalid %select{Memory|Register}1 Interpretation value">;
+def err_hlsl_linalg_matrix_layout_is_not_transposable : Error<
+  "RowMajor and ColumnMajor matrices are not transposable">;
+def err_hlsl_linalg_optimal_matrix_layout_matrix_stride_must_be_zero : Error<
+  "for optimal matrix layout, matrix stride must be zero">;
+def err_hlsl_linalg_matrix_dim_must_not_be_zero: Error<
+  "matrix dimension must not be zero">;
+def err_hlsl_linalg_matrix_layout_invalid : Error<
+  "matrix layout %0 is not valid, must be in the range %1 - %2">;
+def err_hlsl_linalg_function_requires_shader_model_6_9_or_above : Error<
```

Comment:
This should use `warn_hlsl_intrinsic_in_wrong_shader_model` (which is a default-error warning)

Reply:
warnings can be overriden?

Reply:
yes, and for all the availability warnings we allow them to be overridden because they _might_ be valid if the compiler fully optimizes them away.

Reply:
I remembered as I was modifying this per your feedback that I had originally tried to use `warn_hlsl_intrinsic_in_wrong_shader_model` but that error prints out the entry function which is not available here. In my earlier PR I had added check to the `DiagnoseReachableHLSLCall` in SemaHLSL.cpp, but that gets called after these intrinsics are processed. Is there a better place to move this diagnostic?

Reply:
The API availability diagnostics should all be emitted from `DiagnoseReachableHLSLCall`. You can't process it in the intrinsic handling because if it isn't reachable we shouldn't emit the diagnostic.

Reply:
@llvm-beanz
I pushed a change (after your comment) that fixed this by removing this diagnostic, the extra shader model checks, and just relies on the availability diagnostic by adding the `min_sm` attribute to the intrinsics.

This means that other parameter checking diagnostics will fire during parse time, before the shader model is checked.  I think this is what we want.  If so, please resolve, if not, please speak up.  Thanks!